rest/config_test.go

```diff
@@ -799,3 +800,56 @@ func TestParseCommandLineWithConfigContent(t *testing.T) {
 	assert.False(t, guest.Disabled)
 	assert.Equal(t, base.SetFromArray([]string{"*"}), guest.ExplicitChannels)
 }
+
+// This test uses both test_data_bucket and test_indexbucket buckets; these buckets should be available
+// on Couchbase server instance before running this test.
+func TestValidateAndAddDBFromConfig(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("ValidateAndAddDBFromConfig test require Couchbase Bucket")
+	}
+	tmpDir, err := ioutil.TempDir("", "sync_gateway")
+	require.NoError(t, err, "Error whilst creating temporary directory")
+	configFile, err := ioutil.TempFile(tmpDir, "duplicate.bucket.conf")
+	require.NoError(t, err, "Error whilst creating temporary config file")
+
+	content := `{"logging":{"console":{"log_keys":["*"]}},"databases":{"db1":{"server":"couchbase://localhost",
+		"username":"Administrator","password":"password","bucket":"test_data_bucket","users":{"GUEST":{"disabled":false,
+		"admin_channels":["*"]}},"allow_conflicts":false,"revs_limit":20,"num_index_replicas":0},"db2":{"server":
+		"couchbase://localhost","username":"Administrator","password":"password","bucket":"test_data_bucket","users":{
+		"GUEST":{"disabled":false,"admin_channels":["*"]}},"allow_conflicts":false,"revs_limit":20,
+		"num_index_replicas":0},"db3":{"server":"couchbase://localhost","username":"Administrator","password":
+		"password","bucket":"test_indexbucket","users":{"GUEST":{"disabled":false,"admin_channels":["*"]}},
+		"allow_conflicts":false,"revs_limit":20,"num_index_replicas":0}}}`
+
+	_, err = configFile.Write([]byte(content))
+	require.NoError(t, err, "Error whilst writing to temporary config file")
+	require.NoError(t, configFile.Close())
+
+	defer func() {
+		err := os.Remove(configFile.Name())
+		assert.NoError(t, err)
+	}()
+
+	args := []string{"sync_gateway", configFile.Name()}
+	config, err = ParseCommandLine(args, flag.ExitOnError)
+	require.NoError(t, err, "Error whilst parsing commandline options")
+
+	_, err = config.SetupAndValidateLogging()
+	require.NoError(t, err, "Error whilst setting up logging")
+
+	var errors = make([]error, 0)
+	errors = append(errors, config.validate()...)
+	errors = append(errors, config.setupAndValidateDatabases()...)
+	require.Equal(t, 0, len(errors), "Error whilst validating databases")
+
+	sc := NewServerContext(config)
+	warnings := validateAndAddDBFromConfig(sc, config.Databases)
+	warningFormat := "Database %q shares bucket %q with databases %s. " +
+		"This may result in unexpected behaviour if security is not defined consistently."
```

Comment:
Matching on the warning string is not going to be helpful if this message gets changed.

Can we think about having a function return the information in a structured way? E.g:
```go
type sharedBucket struct {
	bucketName string
	dbNames    []string
}

// sharedBuckets returns a list of buckets that are being shared by multiple databases.
func sharedBuckets(uuids map[string][]*db.DatabaseContext) (sharedBuckets []sharedBucket) {
	for _, dbContexts := range uuids {
		if len(dbContexts) > 1 {
			dbNames := make([]string, len(dbContexts))
			for _, dbContext := range dbContexts {
				dbNames = append(dbNames, dbContext.Name)
			}
			sharedBuckets = append(sharedBuckets, sharedBucket{dbContexts[0].Bucket.GetName(), dbNames})
		}
	}
	return sharedBuckets
}
```

The test can do accurate assertions on the results, and the caller can deal with log messages.

Reply:
Makes more sense. Changes are committed to the latest revision.